### PR TITLE
Merge the new refactors

### LINKED ISSUE #669: Suggestions for Twitter module refactor
Apart from the cleanup that @Mira-Alf is doing in the twitter module, just wanted to share some thoughts about twitter module refactoring. I have not understood all of the existing code, but would need help from @Mira-Alf and @PuruVJ to get this into a correct shape. Here is a suggested structure. Please share thoughts based on your understanding of the code, whether it will fit into this kind of abstraction.

image

- `api` (or `service`?) - would handle all the API related content - including the fetch call, the API endpoint, the parameters for the API - and may export only two methods? - one to fetch a specific tweet, and the other to fetch the conversation.
- `scraping` - would hold the puppeteer code, and this folder may have sub folders with utils, etc specific for it.
- `transformations` - would hold any of the methods used to convert the twitter response into a different format (maybe for handlebars consumption)
- `utils`
- `vaildations` - I think there are a few functions used to check the response and generate errors, they could go here.

In the end, this module may expose a couple functions which may be `getTweetThreadFromAPI(threadId)` and `getTweetThreadFromPuppeteer(threadId)` which will return a common format which may be consumed by the handlebars template. Or if we could export a single function which will pick twitter API or puppeteer intelligently, that would be the best.

All of this is from my limited understanding, please discuss below so that we can refine it.

### Update
We had a meeting to discuss about this. Outcome is to move forward with the proposed folder structure. @johnjacobkenny  will take up the refactor in a parallel branch so that the main branch stays stable.
https://www.youtube.com/watch?v=x0v_6QOWAzI

### LINKED ISSUE #676: Create resuable JSDoc types
I've found a way to have make typescript style types in JS using comments only. It will allow us to define the more complex types(Response, tweet object, etc) all in a single file, and the usage would be as simple as

```js

/**
 * @param {TwitterResponse} response
 */
```

Will start work on it ASAP.

Will need reviews by those who have worked with the Twitter API extensively e.g @Mira-Alf

### LINKED ISSUE #698: Twitter module code freeze
Let's focus on porting all extra additions since yesterday evening into the new branch for refactoring.

No more pr for Twitter module into main branch.

@proful @Mira-Alf @PuruVJ @tolgaerdonmez

### LINKED ISSUE #709: Jest test cases

### LINKED ISSUE #710: Twitter refactor part: Is this granularity necessary
I saw a file structure like this in the Twitter refactor:

image

And the contents look like this.
image

My question: Is it really necessary to split them up so much? 
For example, It'll be better to import format directly from date-fns, and it'll convey meaning that it is related to Date formatting only, cuz there might be more formatters later on.

And `const {format} = require('../utils/date')` would be a bit too confusing at first.

And if we're doing this, we better do it right. Let's change the file a bit:
```js
const { format } = require("date-fns");

const formatTimestamp = (timestamp) => format(new Date(timestamp), 'MMM d, yyyy')

module.exports = {
  formatTimestamp,
};

```

What are your thoughts @johnjacobkenny @Mira-Alf  ?

### PR DESCRIPTION
**Pull request by PuruVJ**
Closes #710 
Closes #709 
Closes #698 
Closes #676 
Closes #669 

To @Mira-Alf and @johnjacobkenny for their hard work! 🥂
source

### CODE REVIEW

File: twindle-cli/README.md [hunk 1]
```diff
@@ -28,9 +32,9 @@ It is a wrapper around `console.log` which only prints the input when DEV enviro
 
 There are two options,
 
--   **nodemailer**: Sending attachments to normal mails, send to kindle doesn't get recognized by kindle.
-    Can be found in `utils/send-email.js`
--   **smtp-client**: Currently just sends normal mail, needs further investigation.
-    Can be found in `utils/send-email-smtp.js`
+- **nodemailer**: Sending attachments to normal mails, send to kindle doesn't get recognized by kindle.
+  Can be found in `utils/send-email.js`
+- **smtp-client**: Currently just sends normal mail, needs further investigation.
```
Comment (johnjacobkenny) [inline twindle-cli/README.md:37]: This may be removed.

File: twindle-cli/twitter/api/helpers/fetch.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  if (!token) throw new ApiErrors.TokenNotProvidedError();
+
+  const response = await nodeFetch(url, {
+    method: "GET",
+    headers: { Authorization: `Bearer ${token}` },
+    redirect: "follow",
+  });
+
+  // REVIEW WANTED: What should we do when its not 200? This below is just a workaround for now
+  if (response.status !== 200) throw new ApiErrors.NetworkRequestError();
```
Comment (johnjacobkenny) [inline twindle-cli/twitter/api/helpers/fetch.js:19]: @Mira-Alf any thoughts based on the API?

**Review by johnjacobkenny — APPROVED** [review]
This refactor should give us a good start for us to build upon. Awesome work everyone :partying_face:

File: twindle-cli/README.md (hunk 1 above)
Comment (PuruVJ) [reply twindle-cli/README.md:37]: I'll create an issue

**Comment by johnjacobkenny** [comment]
Annd we are live! Hopefully doesn't break anything lol :rofl:

File: twindle-cli/twitter/api/helpers/fetch.js (hunk 2 above)
Comment (PuruVJ) [reply twindle-cli/twitter/api/helpers/fetch.js:19]: I think we should give more informative error messages. The user will be confused why the error if their Wifi is booming around

Comment (johnjacobkenny) [reply twindle-cli/twitter/api/helpers/fetch.js:19]: @PuruVJ create an issue to discuss

Comment (PuruVJ) [reply twindle-cli/twitter/api/helpers/fetch.js:19]: Sure

**Pull request merged**